File: server/internal/server/objects/units/units/UniversalUnit.go
package units

import (
	"fmt"
	"server/internal/server"
	"server/internal/server/combatDSL"
	"server/internal/server/objects"
	"server/internal/server/objects/tiles"
	"server/internal/server/objects/units/Movement"
	"server/internal/server/objects/units/Skills"
	"server/internal/server/objects/units/sharedInterfaces"
	"server/internal/server/objects/units/units/util"
	"server/pkg/packets"
)

type UniversalUnit struct {
	data            *util.UnitData
	pos             *objects.Vector3
	skills          []Skills.ISkill
	movement        Movement.IMovement
	Tile            *tiles.TerrainInfo
	SkillBuffer     map[int32]combatDSL.CombatAction
	Inflictions     map[int32]InflictAction
	MaxMove         int
	TerrrainService sharedInterfaces.MoveTerrainService
	Service         Skills.GameSkillService
}

func (u *UniversalUnit) NewTurn() {
	for skillID, inflict := range u.Inflictions {
		inflict.Action.Execute(
			combatDSL.CombatContext{
				Caster: u,
				Target: u,
			}, skillID)

		inflict.Turns--
		if inflict.Turns <= 0 {
			delete(u.Inflictions, skillID)
		}
	}
}

func (u *UniversalUnit) RemoveInflict(skillID int32) {
	u.Inflictions[skillID].Action.Invert(skillID)
	delete(u.Inflictions, skillID)
	delete(u.SkillBuffer, skillID)
}

func (u *UniversalUnit) Inflict(skillID int32, action combatDSL.CombatAction, turns int) {
	u.Inflictions[skillID] = InflictAction{
		Action: action,
		Turns:  turns,
	}
}

func (u *UniversalUnit) AddToSkillBuffer(skillID int32, action combatDSL.CombatAction) {
	u.SkillBuffer[skillID] = action
}

func (u *UniversalUnit) RemoveFromSkillBuffer(skillId int32) {
	u.SkillBuffer[skillId].Invert(skillId)
	delete(u.SkillBuffer, skillId)
}

func (u *UniversalUnit) Heal(Amount int64) {
	//TODO implement me
	panic("implement me")
}

func NewUniversalUnit() *UniversalUnit {
	return &UniversalUnit{}
}

func (u *UniversalUnit) NewUnit(data *util.UnitData, client *server.Client, enemyClient *server.Client, service Skills.GameSkillService) {
	fmt.Println(data)
	if data == nil {
		return
	}

	u.data = data

	movementFactory, ok := Movement.MovementRegistry[data.Movement]
	if !ok || movementFactory == nil {
		fmt.Printf("unknown movement type %q for unit id %d", data.Movement, data.UnitID)
	}

	u.movement = movementFactory()
	if u.movement == nil {
		fmt.Printf("movement factory returned nil for movement type %q, unit id %d", data.Movement, data.UnitID)
	}

	u.movement.InitMovement(data.UnitID, client, enemyClient)

	for _, skillName := range u.data.Skills {
		if skillName == "" {
			continue
		}
		skill, ok := Skills.SkillRegistry[skillName]
		if !ok || skill == nil {
			fmt.Printf("unknown skill %q for unit id %d", skillName, data.UnitID)
		}

		skill().Initiate(data.UnitID, client, enemyClient, service)
		u.skills = append(u.skills, skill())
	}
	u.Service = service
	u.TerrrainService = service.World()
	MPAP := 0
	switch data.Speed {
	case util.Fast:
		MPAP = 3
	case util.Normal:
		MPAP = 2
	case util.Slow:
		MPAP = 1
	}
	u.MaxMove = int(data.AP) * MPAP
}
func (u *UniversalUnit) Position() objects.Vector3 {
	return *u.pos
}

func (u *UniversalUnit) SetPosition(pos *objects.Vector3) {
	u.pos = pos
}

func (u *UniversalUnit) Data() *util.UnitData {
	return u.data
}

func (u *UniversalUnit) Skills() []Skills.ISkill {
	return u.skills
}

func (u *UniversalUnit) Attack(unit IUnit) {
	//TODO implement me
	panic("implement me")
}

func (u *UniversalUnit) Movement() *Movement.IMovement {
	return &u.movement
}

func (u *UniversalUnit) Move(path []*packets.HexPositionMessage) bool {
	if u.TerrrainService == nil {
		u.TerrrainService = u.Service.World()
	}
	return u.movement.Move(path, u.TerrrainService, u.MaxMove)
}

func (u *UniversalUnit) Damage(Amount int64) {
	// check if hp is not negative
	if u.data.HP <= 0 {
		// kill unit
		return
	}
	u.data.HP -= Amount
}

File: Client/Objects/Units/Units/UniversalUnit/UniversalUnit.cs
using Godot;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using Packets;

public partial class UniversalUnit : Node3D, IUnit
{
    private const string MeshFolderPath = "res://Objects/Units/Mesh";

    public UnitData Data { get; set; }
    public Node3D TileNode { get; set; }
    [Export] public PackedScene MeshScene, DefaultMeshScene;
    public DefaultUnit Mesh;
    public Color Color;
    public Dictionary<string, ISkill> Skills { get; set; }
    public IMovement Movement { get; set; }
    public Vector2I PositionI { get; set; }
    public Dictionary<int, ICombatAction> SkillBuffer { get; set; } = new(); // <skillId, ICombatAction>
    public Dictionary<int, InflictAction> Inflictions { get; set; } = new(); // <skillId, InflictAction>

    #region Init

    public override void _Ready()
    {
        InitUnit();
    }

    public virtual void InitUnit()
    {
        if (Data == null)
        {
            GD.PrintErr("Unit data is null!");
            return;
        }

        if (MeshScene == null)
        {
            FindMesh(Data.UnitName);
        }

        Mesh = MeshScene?.Instantiate<DefaultUnit>();
        BuildNamePlate();
        Mesh?.ChangeColor(Color);
        AddChild(Mesh);


        FindSkills(Data.Skills);
        FindMovement(Data.Movement);
        Movement = new DefaultMovement();
        Movement.InitMovement(this, this);
        Skills?.Values.ToList().ForEach(a => a.Init(this, this, null)); // skillData

    }

    private void BuildNamePlate()
    {
        Mesh.NamePlate.Text = Data.UnitName;
    }

    private void FindMesh(string meshName)
    {
        if (string.IsNullOrWhiteSpace(meshName))
        {
            GD.PrintErr("Mesh name is empty.");
            MeshScene = DefaultMeshScene;
            return;
        }

        string scenePath = FindScenePathRecursive(MeshFolderPath, meshName);

        if (string.IsNullOrEmpty(scenePath))
        {
            GD.PrintErr($"Mesh scene '{meshName}' was not found in '{MeshFolderPath}'.");
            MeshScene = DefaultMeshScene;
            return;
        }

        MeshScene = ResourceLoader.Load<PackedScene>(scenePath);

        if (MeshScene == null)
        {
            GD.PrintErr($"Failed to load mesh scene: {scenePath}");
        }
    }

    private void FindSkills(List<string> skills)
    {
        if (skills == null || skills.Count <= 0 || (skills.Count == 1 && skills[0] == ""))
        {
            return;
        }

        Skills = new();
        foreach (var skillName in skills)
        {
            var skillType = Assembly.GetExecutingAssembly().GetTypes().Where(t =>
                    typeof(ISkill).IsAssignableFrom(t) && !t.IsInterface && !t.IsAbstract && t.Name == skillName)
                .ToList()[0];
            var skill = (ISkill)Activator.CreateInstance(skillType);
            Skills.Add(skillName, skill);
        }
    }

    private void FindMovement(string movementName)
    {
        var movementType = Assembly.GetExecutingAssembly().GetTypes().Where(t =>
                typeof(IMovement).IsAssignableFrom(t) && !t.IsInterface && !t.IsAbstract && t.Name == movementName)
            .ToList()[0];
        Movement = (IMovement)Activator.CreateInstance(movementType);
    }

    private string FindScenePathRecursive(string folderPath, string meshName)
    {
        using DirAccess dir = DirAccess.Open(folderPath);

        if (dir == null)
        {
            GD.PrintErr($"Could not open directory: {folderPath}");
            return null;
        }

        foreach (string fileName in dir.GetFiles())
        {
            string extension = fileName.GetExtension();

            if (extension != "tscn" && extension != "scn")
                continue;

            string fileBaseName = fileName.GetBaseName();

            if (!string.Equals(fileBaseName, meshName, StringComparison.OrdinalIgnoreCase))
                continue;

            return $"{folderPath}/{fileName}";
        }

        foreach (string directoryName in dir.GetDirectories())
        {
            if (directoryName.StartsWith("."))
                continue;

            string result = FindScenePathRecursive($"{folderPath}/{directoryName}", meshName);

            if (!string.IsNullOrEmpty(result))
                return result;
        }

        return null;
    }

    #endregion

    public void NewTurn()
    {
        // process inflictions
        foreach (var inflict in Inflictions)
        {
            inflict.Value.Action.Execute(new CombatContext(this, this), inflict.Key);
            inflict.Value.Turns--;
            if (inflict.Value.Turns <= 0)
            {
                Inflictions.Remove(inflict.Key);
            }
        }
    }

    public virtual void AddToSkillBuffer(int skillId, ICombatAction action)
    {
        SkillBuffer.Add(skillId, action);
    }

    public void RemoveFromSkillBuffer(int skillId)
    {
        SkillBuffer[skillId].Invert(skillId);
        SkillBuffer.Remove(skillId);
    }

    public virtual void Move(TerrainInfo fromPos, TerrainInfo toPos, bool message = false)
    {
        var path = PathFinding.FindCheapestPath(fromPos, toPos);
        Movement.Move(path, message);
    }
    public virtual void Skill(Vector2I position, string skillName)
    {
        var skill = Skills[skillName];
        skill.Use(position);
    }

    public void Inflict(int skillId, ICombatAction action, int Turns)
    {
        Inflictions.Add(skillId, new InflictAction() { Action = action, Turns = Turns });
    }

    public void RemoveInflict(int skillId)
    {
        throw new NotImplementedException();
    }

    public void Damage(float amount)
    {
        Data.HP -= (long)amount;
    }
    public void Heal(float amount)
    {
        Data.HP += (long)amount;
    }
}